### PR TITLE
Avoid more cases of buffer overread.

### PR DESCRIPTION
**Pull request by toregge**
@vekterli : please review

### CODE REVIEW

**Review by vekterli — APPROVED** [review]
:+1:

File: document/src/vespa/document/datatype/mapdatatype.cpp [hunk 1]
```diff
@@ -63,27 +69,27 @@ MapDataType::buildFieldPathImpl(FieldPath & path, const DataType &dataType,
         std::string_view rest = remainFieldName;
         vespalib::string keyValue = FieldPathEntry::parseKey(rest);
 
-        valueType.buildFieldPath(path, (rest[0] == '.') ? rest.substr(1) : rest);
+        valueType.buildFieldPath(path, (!rest.empty() && rest[0] == '.') ? rest.substr(1) : rest);
 
-        if (remainFieldName[1] == '$') {
+        if (remainFieldName.size() > 1 && remainFieldName[1] == '$') {
```
Comment (baldersheim) [inline document/src/vespa/document/datatype/mapdatatype.cpp:74]: Perhaps add utility function similar too

bool safe_cmp(const string & s, char c, size_t pos) {
    return (s.size() > pos). && (s[pos] == c);
}

File: document/src/vespa/document/base/fieldpath.cpp [hunk 2]
```diff
@@ -141,9 +142,9 @@ FieldPathEntry::parseKey(std::string_view & key)
     vespalib::string v;
     const char *c = key.data();
     const char *e = c + key.size();
-    for(;(c < e) && isspace(c[0]); c++);
+    for(;(c < e) && std::isspace(static_cast<unsigned char>(c[0])); c++);
     if ((c < e) && (c[0] == '{')) {
-        for(c++;(c < e) && isspace(c[0]); c++);
+        for(c++;(c < e) && std::isspace(static_cast<unsigned char>(c[0])); c++);
```
Comment (baldersheim) [inline document/src/vespa/document/base/fieldpath.cpp:147]: perhaps an utility method similar to

const char * eatWhite(const char * s, const char * e) {
    for(;(c < e) && std::isspace(static_cast<unsigned char>(c[0])); c++);
    return c;
}

**Review by baldersheim — APPROVED** [review]
Looks good, but perhaps take the opportunity to improve readability.

**Pull request merged**